### ISSUE TITLE
Active Menu Highlight not work on Blazor WebAssembly application

### ISSUE THREAD

**Issue by sounj142** [ShaunCurtis/CEC.Routing#4]
Hi @ShaunCurtis,

I'm implementing a Blazor WebAssembly application, I use your library to prevent user navigate to another page without save changes. Everything looks good except the highlight for current menu.

Here is the screenshot. I converted your sample - CEC.RoutingSample - to blazor wasm to re-check the problem.

![image](https://user-images.githubusercontent.com/12889119/90652319-2563e000-e268-11ea-8c69-dc0a062eab76.png)

As you can see, the current menu is "Add Forecast", I input something to form, click "Counter". Library successfully prevents browser navigate to Counter menu, but it highlights "Counter" menu, an unexpected behavior

**Comment by ShaunCurtis** [comment]
Hi Hoang Luong,
I do mostly Blazor Server development, so haven't looked at the routing in Blazor WASM.  I have a couple of issues to fix and roll out shortly, so will look at this when I do them.  It works fine in Server mode, the highlighting is set by the NavLink control which uses the NavigationManager.   In the current implementation the router does a second "dummy" trip through the NavigationManager to set the NavigationManager page URL back to the original page.  I can see it's set correctly on your page.  There's two Navigation Managers - WebAssemblyNavigationManager for WASM and RemoteNavigationManager for Blazor Server  - so I'm guessing the problem lies there.

**Comment by ShaunCurtis** [comment]
Hi Hoang Luong,

Not forgotten.  I've tracked the problem down to some issues with navigation manager, and have a work around ready to release tomorrow.  To keep it short, when RecordRouter detects a navigation event it needs to cancel , it reruns the Navigation event to point NavigationManager back to the original page.  In this case clicked on Counter causes the navigation manager to  trigger a LocationChanged event.  The RecordRouter - being an early registerer on the event gets the event early.  It runs it's event code which ends with a new NavigateTo event.  This triggers another LocationChanged event.  The problem for NavLink is there's some blocking code somewhere along the line that blocks the thread, and the second LocationChnaged event completes before the first one.  The result is that NavLink gets the second event before the first - and thus the url is to counter.  I need to dig into the MS code to figure out who the offending blocker is, but the workaround is:
1.  An async method in the router as follows:
        private async Task ReNavigate()
        {
            await Task.Delay(250);
            var url = this.RouterSessionService.RouteUrl;
            this.NavigationManager.NavigateTo(this.RouterSessionService.RouteUrl);
        }
2. Marking the OnLocationChanged event in the router as Async:

        private async void OnLocationChanged(object sender, LocationChangedEventArgs args)

3. changing the code at the end of the OnLocationChanged event in the router to:
                    .....
                    this.RouterSessionService.NavigationCancelledUrl = this.NavigationManager.Uri;
                    await ReNavigate();
                }

This yields control on the OnLocationChanged event on the router and waits a 250 msecs before re-navigating.

You should see a new version of the Nuget package and the GitHub repo updated by Sunday.

**Comment by sounj142** [comment]
Hi Shaun Curtis,

Thank you for your support,  you’re so awesome 👍 
I think your work around way will work, will looking forward to the new update

Currently, my workaround is not beauty, I had to remove all built-in NavMenu components and control html directly because the built-in ones depend on OnLocationChanged and didn't work correctly.

```csharp
<a href="@Constants.CandidatesPath" class="b-bar-dropdown-item @ClassForMenu(Constants.CandidatesPath)">
	Candidates
</a>
```
In the code behind, I still register LocationChanged event but I totally ignore the "wrong" parameter LocationChangedEventArgs, all the thing I do is call StateHasChanged to say "Hey, something change", and let ClassForMenu method decide whether to highlight or not
```csharp
private void OnLocationChanged(object sender, LocationChangedEventArgs e)
{
	StateHasChanged();
}

private string ClassForMenu(string prefixPath)
{
	bool isCurrentPath = navigationManager.Uri.StartsWith(navigationManager.ToAbsoluteUri(prefixPath).ToString(), StringComparison.OrdinalIgnoreCase);
	return isCurrentPath ? "active" : null;
}
```

**Comment by ShaunCurtis** [comment]
Thanks.

Code is now all published here on GitHub and new 1.2 release is available at Nuget.

WASM applications are single threaded - no choice in the browser, that's what you currently get.  So Javascript and WASM run on the same thread, and as most of the code is synchronous then ....
